### PR TITLE
Deduplicate tensor pad-to-size code

### PR DESCRIPTION
**Pull request by akx**
```
ast-grep run --pattern '($A - $B % $A) % $A' -l python
```
revealed a bunch of places that could use some reuse. 😄

### CODE REVIEW

**Comment by RunDevelopment** [comment]
What you essentially did in this PR is to factor out common code between architectures. This raises a general question: should we refactor arch code? 

IMO, we should not. For 3 reasons:
1. It doesn't scale. As we add more and more architectures, is this something you/we will continue doing? I don't want to force first-time contributors adding an arch to refactor highly sophisticated network architectures, so who will do it?
1. There is no benefit (that I can see). In a normal code base, refactoring and principles like DRY are doing to keep the maintainable = keep it easy to change. But arch code doesn't change. We don't continue to develop architectures.
1. Architectures are currently almost fully independent of each other, and I would like to keep them decoupled. E.g. the SwinIR derivatives share many classes with SwinIR. Should we make those architectures depend on SwinIR to reduce code duplication and couple them in the progress?

What are the reasons for why we should generally do this?

**Comment by joeyballentine** [comment]
I generally agree with that statement in that we shouldn't try to change arch code.

HOWEVER: The more arch code we can dedupe, the smaller spandrel will end up being. It might not be a significant amount yet, but as we add more arches that's gonna add up eventually. But I still don't know if we'd save enough to make it worth it.

**Comment by RunDevelopment** [comment]
What is "smaller" here? I'm assuming wheels are compressed, so it shouldn't affect distribution size.

**Comment by joeyballentine** [comment]
Well, each release of spandrel has made it grow in size. It's still in the kilobytes but I keep having to change the estimated size in chaiNNer's dependency entry for it

**Comment by RunDevelopment** [comment]
Again, would this actually help with the distribution size? (I just checked) wheels are compressed, so would removing code duplication actually help?

**Comment by joeyballentine** [comment]
I guess not then

**Comment by akx** [comment]
> 1. It doesn't scale. As we add more and more architectures, is this something you/we will continue doing? I don't want to force first-time contributors adding an arch to refactor highly sophisticated network architectures, so who will do it?

I wouldn't say it needs to be forced at all? An arch could be implemented as-is and cleaned up over time.

> 2. There is no benefit (that I can see). In a normal code base, refactoring and principles like DRY are doing to keep the maintainable = keep it easy to change. 

[DRY/SSOT](https://en.wikipedia.org/wiki/Single_source_of_truth) cleanups like this make the codebase easier to read and debug, IMO – sure, you can learn to recognize the two-`$C = ($A - $B % $A) % $A`-followed-by-`F.pad()` pattern, but I think `pad_to_multiple` is more understandable for e.g. someone who needs to debug the code.

> But arch code doesn't change. We don't continue to develop architectures.

Which is why it's, IMO, even more fine to do cleanups like this :)

> 3. Architectures are currently almost fully independent of each other, and I would like to keep them decoupled.

There are some block helpers already (I guess that's the "almost" you refer to) that are shared between architectures.

> E.g. the SwinIR derivatives share many classes with SwinIR. Should we make those architectures depend on SwinIR to reduce code duplication and couple them in the progress?

Since the arch code is set in stone, I don't see why not (since there is a robust test suite to catch regressions).

Looking at PyCharm's "Locate Duplicates", there's e.g.
`spandrel.architectures.DDColor.arch.position_encoding.PositionEmbeddingSine` that's character by character the exact same code as `spandrel.architectures.CodeFormer.arch.codeformer.PositionEmbeddingSine`, with the same comment.  (Of course, it looks like CodeFormer's `PositionEmbeddingSine` is never even referenced, so there's that...)

Or `MMRRDBNet` and `MMRRDBNet_decouple`, where the differences are:

* a comment `# 512 is real!!!!!` vs `#  512 is reall`
* an internal `with torch.no_grad():`
* a `.detach()` call that looks inconsequential to me

> Again, would this actually help with the distribution size? (I just checked) wheels are compressed, so would removing code duplication actually help?

It would probably make some difference, considering ZIPs (which is what wheels are) compress each file separately, so they wouldn't share zlib compressor state.

**Comment by joeyballentine** [comment]
You make a good case about being able to make those changes confidently since we have tests. 

I think it might be a good idea to at least attempt it

**Comment by RunDevelopment** [comment]
Alright, if you two want to do it, then do it. I just want to say, that seems like a lot of work, for little benefit and the risk of introducing regressions.

> There are some block helpers already (I guess that's the "almost" you refer to) that are shared between architectures.

Yup. Joey reused some ESRGAN blocks, I think, and we vendored timm.

> since there is a robust test suite to catch regressions

Be careful. We just test a handful of pretrained models and hyperparameter configurations for each arch (and sometimes not even pretrains). E.g. we just recently found out that ESRGAN+ models didn't work because of an oversight.

> considering ZIPs

Yeah, in that case it would reduce dist size.

**Comment by joeyballentine** [comment]
> Yup. Joey reused some ESRGAN blocks, I think

Just to clarify this: those blocks were copied from the original implementation within BasicSR/the original ESRGAN repo. I didn't manually dedupe anything myself.

As for this PR, now that there's been some time, do we feel any differently about if we should merge this or not? I could really go either way on it

**Comment by RunDevelopment** [comment]
Honestly, might as well. I'm somewhat against the idea of taking deduplication between archs too far, but this PR seems fine. Especially, since we might use `pad_to_multiple` in other places in the future as well. E.g. we might want to handle size requirements in the call API (automatically pad + unpad for some arch).

**Comment by akx** [comment]
> @akx Could you please resolve the conflicts? (I don't even know what conflicts it has, but whatever...)

Done (simple rebase did it). I think it was a conflict that could have been a problem for GitHub's naive merge, but isn't for a rebase.

> Did you uncheck the "Allow edits from maintainers" thingy when you made the PR? 

I tend to do that in general, because I don't mind personally rebasing and fixing my PRs when asked to 😄

**Comment by akx** [comment]
Ah, `DAT.py` has one more `'($A - $B % $A) % $A'`, gonna add that too.

**Comment by RunDevelopment** [comment]
Thank you @akx!
